pkg/virtual/apiexport/authorizer/maximal_permission_policy.go

```diff
@@ -0,0 +1,192 @@
+		},
+		getAPIExportsByIdentity: func(identityHash string) ([]*apisv1alpha1.APIExport, error) {
+			exports, err := apiExportIndexer.ByIndex(indexers.APIExportByIdentity, identityHash)
+			if err != nil {
+				return nil, err
+			}
+			var results []*apisv1alpha1.APIExport
+			for _, obj := range exports {
+				results = append(results, obj.(*apisv1alpha1.APIExport))
+			}
+			return results, nil
+		},
+		newDeepSARAuthorizer: func(clusterName logicalcluster.Name) (authorizer.Authorizer, error) {
+			return delegated.NewDelegatedAuthorizer(clusterName, deepSARClient)
+		},
+	}
+}
+
+func (a *maximalPermissionAuthorizer) Authorize(ctx context.Context, attr authorizer.Attributes) (authorizer.Decision, string, error) {
+	apiDomainKey := dynamiccontext.APIDomainKeyFrom(ctx)
+	parts := strings.Split(string(apiDomainKey), "/")
+	if len(parts) < 2 {
+		err := fmt.Errorf("invalid API domain key: %q", apiDomainKey)
+		return authorizer.DecisionNoOpinion, VirtualAPIExportNotPermittedReason, err
+	}
+
+	claimingAPIExportCluster := parts[0]
+	claimingAPIExportName := parts[1]
+
+	claimingAPIExport, err := a.getAPIExport(claimingAPIExportCluster, claimingAPIExportName)
+	if err != nil && kerrors.IsNotFound(err) {
+		return authorizer.DecisionNoOpinion, VirtualAPIExportNotPermittedReason, err
+	}
+
+	claimedIdentityHash, found := getClaimedIdentity(claimingAPIExport, attr)
+	if !found {
+		// it's a resource in the claiming API export, hence unclaimed
+		return authorizer.DecisionAllow, notEnforcedMaximumPermissionPolicyReason, nil
+	}
+	if claimedIdentityHash == "" {
+		// it's a native k8s resource (secret, configmap, ...), or a system kcp CRD resource (apis.kcp.dev)
+		// For neither case a maximum permission policy can exist.
+		return authorizer.DecisionAllow, notEnforcedMaximumPermissionPolicyReason, nil
+	}
+
+	claimedAPIExports, err := a.getAPIExportsByIdentity(claimedIdentityHash)
+	if err != nil {
+		return authorizer.DecisionNoOpinion, VirtualAPIExportNotPermittedReason, err
+	}
+
+	if len(claimedAPIExports) == 0 {
+		// a claimed identity hash exists but not API export can be found referring to it (potentially eventually consistent).
+		// In this case be safe and deny the request, forcing the caller to retry.
+		return authorizer.DecisionDeny, VirtualAPIExportNotPermittedReason, nil
+	}
+
+	// multiple claimed API exports can share the same identity hash (even in different workspaces).
+	// All maximum permission policies must grant access because at this point no deterministic API export can be picked
+	// (especially in the wildcard case).
+	for _, claimedAPIExport := range claimedAPIExports {
+		if claimedAPIExport.Spec.MaximalPermissionPolicy == nil {
+			continue
+		}
+
+		if claimedAPIExport.Spec.MaximalPermissionPolicy.Local == nil {
+			continue
+		}
+
+		prefixedAttr := deepCopyAttributes(attr)
+		userInfo := prefixedAttr.User.(*user.DefaultInfo)
```

Comment:
My initial reaction here was to ask why we're type asserting to `*user.DefaultInfo`. I then realized it was so we could modify the user and groups to prepend the maximal permission policy prefix. I see that deepCopyAttributes is only used in this file and added in this PR. Would it be possible to make some edits so maybe instead of doing this in two steps (deep copy, then adjust user/groups), we just do it in one, and we can avoid the type assertion?

Reply:
oh sure thing 👍 this is pure copy+pasta from somewhere else, let me try to enhance this.

Reply:
done